README.md
# callistor
<!-- badges: start -->
[![R-CMD-check](https://github.com/OakCityLabs/callisto-r/workflows/R-CMD-check/badge.svg)](https://github.com/OakCityLabs/callisto-r/actions)
<!-- badges: end -->

Provides functions for an R jupyter kernel running on a Callisto Jupyter D1 server to communicate with the D1 server via json. For example, callistor provides methods to get json representations of variables defined in the kernel


NEWS.md
# callistor 0.0.1
* Initial release, with functions to get json representations of variables in a provided environment


tests/testthat.R
library(testthat)
library(callistor)

test_check("callistor")


tests/testthat/test_vars_manager.R
test_that("human_bytes returns bytes", {
    expect_equal(human_bytes(1023), "1023 Bytes")
    expect_equal(human_bytes(1024), "1.00 KB")
    expect_equal(human_bytes(1024), "1.00 KB")
    expect_equal(human_bytes(1048575), "1024.00 KB")
    expect_equal(human_bytes(1048576), "1.00 MB")
    expect_equal(human_bytes(1073741823), "1024.00 MB")
    expect_equal(human_bytes(1073741824), "1.00 GB")
})

test_that("format_vars single element character vector", {
    string1 <- "efs"
    vars <- format_vars(environment())
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "string1")
    expect_equal(parsed_vars[[1]]$type, "character")
    expect_equal(parsed_vars[[1]]$abbreviated, FALSE)
    expect_equal(parsed_vars[[1]]$summary, "efs")
    expect_equal(parsed_vars[[1]]$value$single_value, "efs")
})

test_that("format_vars single element numeric vector", {
    numeric1 <- 342.3
    vars <- format_vars(environment())
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "numeric1")
    expect_equal(parsed_vars[[1]]$type, "numeric")
    expect_equal(parsed_vars[[1]]$abbreviated, FALSE)
    expect_equal(parsed_vars[[1]]$summary, "342.3")
    expect_equal(parsed_vars[[1]]$value$single_value, "342.3")
})

test_that("format_vars single element integer vector", {
    integer1 <- 9
    vars <- format_vars(environment())
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "integer1")
    expect_equal(parsed_vars[[1]]$type, "numeric")
    expect_equal(parsed_vars[[1]]$abbreviated, FALSE)
    expect_equal(parsed_vars[[1]]$summary, "9")
    expect_equal(parsed_vars[[1]]$value$single_value, "9")
})

test_that("format_vars single element logical vector", {
    logical1 <- TRUE
    vars <- format_vars(environment())
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "logical1")
    expect_equal(parsed_vars[[1]]$type, "logical")
    expect_equal(parsed_vars[[1]]$abbreviated, FALSE)
    expect_equal(parsed_vars[[1]]$summary, "TRUE")
    expect_equal(parsed_vars[[1]]$value$single_value, "TRUE")
})

test_that("format_vars single element complex vector", {
    complex1 <- 1 + 2i 
    vars <- format_vars(environment())
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "complex1")
    expect_equal(parsed_vars[[1]]$type, "complex")
    expect_equal(parsed_vars[[1]]$abbreviated, FALSE)
    expect_equal(parsed_vars[[1]]$summary, "1+2i")
    expect_equal(parsed_vars[[1]]$value$single_value, "1+2i")
})

test_that("format_vars multi element vector", {
    vector1 <- c(4, 3)
    vars <- format_vars(environment())
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "vector1")
    expect_equal(parsed_vars[[1]]$type, "numeric")
    expect_equal(parsed_vars[[1]]$abbreviated, FALSE)
    expect_equal(parsed_vars[[1]]$summary, "Length: 2")
    expect_equal(parsed_vars[[1]]$value$multi_value$column_count, 1)
    expect_equal(parsed_vars[[1]]$value$multi_value$row_count, 2)
    expect_equal(parsed_vars[[1]]$value$multi_value$column_names, c("vector1"))
    expect_equal(parsed_vars[[1]]$value$multi_value$row_names, c("1", "2"))
    expect_equal(parsed_vars[[1]]$value$multi_value$data, c("4", "3"))
})

test_that("format_vars multi element list", {
    list1 <- list(4, 3)
    vars <- format_vars(environment())
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "list1")
    expect_equal(parsed_vars[[1]]$type, "list")
    expect_equal(parsed_vars[[1]]$abbreviated, FALSE)
    expect_equal(parsed_vars[[1]]$summary, "Length: 2")
    expect_equal(parsed_vars[[1]]$value$multi_value$column_count, 1)
    expect_equal(parsed_vars[[1]]$value$multi_value$row_count, 2)
    expect_equal(parsed_vars[[1]]$value$multi_value$column_names, c("list1"))
    expect_equal(parsed_vars[[1]]$value$multi_value$row_names, c("1", "2"))
    expect_equal(parsed_vars[[1]]$value$multi_value$data, c("4", "3"))
})

test_that("format_vars multi element list abbreviated", {
    list1 <- as.list(300:1)
    vars <- format_vars(environment())
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "list1")
    expect_equal(parsed_vars[[1]]$type, "list")
    expect_equal(parsed_vars[[1]]$abbreviated, TRUE)
    expect_equal(parsed_vars[[1]]$summary, "Length: 300")
    expect_equal(parsed_vars[[1]]$value$multi_value$column_count, 1)
    expect_equal(parsed_vars[[1]]$value$multi_value$row_count, 300)
    expect_equal(parsed_vars[[1]]$value$multi_value$column_names, c("list1"))
    expect_equal(parsed_vars[[1]]$value$multi_value$row_names, as.character(1:50))
    expect_equal(parsed_vars[[1]]$value$multi_value$data, as.character(300:251))
})

test_that("format_vars single element named list", {
    list1 <- list(cool="kat")
    vars <- format_vars(environment())
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "list1")
    expect_equal(parsed_vars[[1]]$type, "list")
    expect_equal(parsed_vars[[1]]$abbreviated, FALSE)
    expect_equal(parsed_vars[[1]]$summary, "Length: 1")
    expect_equal(parsed_vars[[1]]$value$multi_value$column_count, 1)
    expect_equal(parsed_vars[[1]]$value$multi_value$row_count, 1)
    expect_equal(parsed_vars[[1]]$value$multi_value$column_names, c("list1"))
    expect_equal(parsed_vars[[1]]$value$multi_value$row_names, c("cool"))
    expect_equal(parsed_vars[[1]]$value$multi_value$data, c("kat"))
})

test_that("format_vars multiple element named list abbreviated", {
    list1 <- list()
    for (i in 1:300) {
        list1[toString(i)] = i
    }
    rm(i)
    vars <- format_vars(environment())
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "list1")
    expect_equal(parsed_vars[[1]]$type, "list")
    expect_equal(parsed_vars[[1]]$abbreviated, TRUE)
    expect_equal(parsed_vars[[1]]$summary, "Length: 300")
    expect_equal(parsed_vars[[1]]$value$multi_value$column_count, 1)
    expect_equal(parsed_vars[[1]]$value$multi_value$row_count, 300)
    expect_equal(parsed_vars[[1]]$value$multi_value$column_names, c("list1"))
    expect_equal(parsed_vars[[1]]$value$multi_value$row_names, as.character(1:50))
    expect_equal(parsed_vars[[1]]$value$multi_value$data, as.character(1:50))
})

test_that("format_vars matrix", {
    matrix1 <- matrix(nrow=4, ncol=7)
    for (i in 1:4) {
        for (j in 1:7) {
            matrix1[i,j] = i * 4 + j
        }
    }
    rm(i)
    rm(j)
    vars <- format_vars(environment())
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "matrix1")
    expect_equal(grepl("matrix", parsed_vars[[1]]$type), TRUE)
    expect_equal(parsed_vars[[1]]$abbreviated, FALSE)
    expect_equal(startsWith(parsed_vars[[1]]$summary, "Size: 4x7 Memory: "), TRUE)
    expect_equal(parsed_vars[[1]]$value$multi_value$column_count, 7)
    expect_equal(parsed_vars[[1]]$value$multi_value$row_count, 4)
    expect_equal(parsed_vars[[1]]$value$multi_value$column_names, as.character(1:7))
    expect_equal(parsed_vars[[1]]$value$multi_value$row_names, as.character(1:4))
    mdata <- c()
    for (i in 1:4) {
        mdata[[i]] = as.character(i * 4 + 1:7)
    }
    expect_equal(parsed_vars[[1]]$value$multi_value$data, mdata)
})

test_that("format_vars matrix abbreviated", {
    matrix1 <- matrix(nrow=80, ncol=7)
    for (i in 1:80) {
        for (j in 1:7) {
            matrix1[i,j] = i * 4 + j
        }
    }
    rm(i)
    rm(j)
    vars <- format_vars(environment())
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "matrix1")
    expect_equal(grepl("matrix", parsed_vars[[1]]$type), TRUE)
    expect_equal(parsed_vars[[1]]$abbreviated, TRUE)
    expect_equal(startsWith(parsed_vars[[1]]$summary, "Size: 80x7 Memory: "), TRUE)
    expect_equal(parsed_vars[[1]]$value$multi_value$column_count, 7)
    expect_equal(parsed_vars[[1]]$value$multi_value$row_count, 80)
    expect_equal(parsed_vars[[1]]$value$multi_value$column_names, as.character(1:7))
    expect_equal(parsed_vars[[1]]$value$multi_value$row_names, as.character(1:8))
    mdata <- c()
    for (i in 1:8) {
        mdata[[i]] = as.character(i * 4 + 1:7)
    }
    expect_equal(parsed_vars[[1]]$value$multi_value$data, mdata)
})

test_that("format_vars data.frame abbreviated", {
    dataframe1 <- read.csv("iris.csv", colClasses=c("numeric", "numeric", "numeric", "numeric", "factor"))
    vars <- format_vars(environment(), 33)
    parsed_vars = rjson::fromJSON(vars)
    expect_equal(length(parsed_vars), 1)
    expect_equal(parsed_vars[[1]]$name, "dataframe1")
    expect_equal(parsed_vars[[1]]$type, "data.frame")
    expect_equal(parsed_vars[[1]]$abbreviated, TRUE)
    expect_equal(startsWith(parsed_vars[[1]]$summary, "Size: 150x5 Memory: "), TRUE)
    expect_equal(parsed_vars[[1]]$value$multi_value$column_count, 5)
    expect_equal(parsed_vars[[1]]$value$multi_value$row_count, 150)
    expect_equal(
        parsed_vars[[1]]$value$multi_value$column_names, 
        c(
            "sepallength (numeric)",
            "sepalwidth (numeric)",
            "petallength (numeric)",
            "petalwidth (numeric)",
            "class (factor)"
        )
    )
    expect_equal(parsed_vars[[1]]$value$multi_value$row_names, as.character(1:7))
    expect_equal(length(parsed_vars[[1]]$value$multi_value$data), 7)
})

test_that("format_var multi element list long but not abbreviated", {
    list1 <- as.list(300:1)
    vars <- format_var(environment(), "list1", NULL)
    parsed_var = rjson::fromJSON(vars)
    expect_equal(parsed_var$name, "list1")
    expect_equal(parsed_var$type, "list")
    expect_equal(parsed_var$abbreviated, FALSE)
    expect_equal(parsed_var$summary, "Length: 300")
    expect_equal(parsed_var$value$multi_value$column_count, 1)
    expect_equal(parsed_var$value$multi_value$row_count, 300)
    expect_equal(parsed_var$value$multi_value$column_names, c("list1"))
    expect_equal(parsed_var$value$multi_value$row_names, as.character(1:300))
    expect_equal(parsed_var$value$multi_value$data, as.character(300:1))
})

test_that("format_var multi element named list long but not abbreviated", {
    list1 <- list()
    for (i in 1:300) {
        list1[toString(i)] = i
    }
    rm(i)
    vars <- format_var(environment(), "list1", NULL)
    parsed_var = rjson::fromJSON(vars)
    expect_equal(parsed_var$name, "list1")
    expect_equal(parsed_var$type, "list")
    expect_equal(parsed_var$abbreviated, FALSE)
    expect_equal(parsed_var$summary, "Length: 300")
    expect_equal(parsed_var$value$multi_value$column_count, 1)
    expect_equal(parsed_var$value$multi_value$row_count, 300)
    expect_equal(parsed_var$value$multi_value$column_names, c("list1"))
    row_names = as.character(1:300)
    expect_equal(parsed_var$value$multi_value$row_names, row_names)
    expect_equal(parsed_var$value$multi_value$data, as.character(1:300))
})

test_that("format_var matrix long but not abbreviated", {
    matrix1 <- matrix(nrow=80, ncol=7)
    for (i in 1:80) {
        for (j in 1:7) {
            matrix1[i,j] = i * 4 + j
        }
    }
    rm(i)
    rm(j)
    vars <- format_var(environment(), "matrix1", NULL)
    parsed_var = rjson::fromJSON(vars)
    expect_equal(parsed_var$name, "matrix1")
    expect_equal(grepl("matrix", parsed_var$type), TRUE)
    expect_equal(parsed_var$abbreviated, FALSE)
    expect_equal(startsWith(parsed_var$summary, "Size: 80x7 Memory: "), TRUE)
    expect_equal(parsed_var$value$multi_value$column_count, 7)
    expect_equal(parsed_var$value$multi_value$row_count, 80)
    expect_equal(parsed_var$value$multi_value$column_names, as.character(1:7))
    expect_equal(parsed_var$value$multi_value$row_names, as.character(1:80))
    mdata <- c()
    for (i in 1:80) {
        mdata[[i]] = as.character(i * 4 + 1:7)
    }
    expect_equal(parsed_var$value$multi_value$data, mdata)
})

test_that("format_var data.frame long but not abbreviated", {
    dataframe1 <- read.csv("iris.csv", colClasses=c("numeric", "numeric", "numeric", "numeric", "character"))
    vars <- format_var(environment(), "dataframe1", NULL)
    parsed_var = rjson::fromJSON(vars)
    expect_equal(parsed_var$name, "dataframe1")
    expect_equal(parsed_var$type, "data.frame")
    expect_equal(parsed_var$abbreviated, FALSE)
    expect_equal(startsWith(parsed_var$summary, "Size: 150x5 Memory: "), TRUE)
    expect_equal(parsed_var$value$multi_value$column_count, 5)
    expect_equal(parsed_var$value$multi_value$row_count, 150)
    expect_equal(
        parsed_var$value$multi_value$column_names,
        c(
            "sepallength (numeric)",
            "sepalwidth (numeric)",
            "petallength (numeric)",
            "petalwidth (numeric)",
            "class (character)"
        )
    )
    expect_equal(parsed_var$value$multi_value$row_names, as.character(1:150))
    expect_equal(length(parsed_var$value$multi_value$data), 150)
})

test_that("format_var error", {
    expect_error(format_var(environment(), "fakevar", NULL), "object 'fakevar' not found")
})

test_that("create_exception_var returns formatted error", {
    err_resp <- tryCatch({
        stop("something exploded")
    }, error=function(e) {
        return(e)
    })
    error_var = create_exception_var(err_resp)
    expect_equal(error_var$name, "Introspection Error")
    expect_equal(error_var$type, "simpleError, error, condition")
    expect_equal(error_var$abbreviated, FALSE)
    expect_equal(error_var$summary, "Error in doTryCatch(return(expr), name, parentenv, handler): something exploded\n")
    expect_equal(error_var$value$multi_value$column_count, 1)
    expect_equal(error_var$value$multi_value$row_count > 1, TRUE)
    expect_equal(error_var$value$multi_value$column_names, list("Traceback"))
    expect_equal(length(error_var$value$multi_value$row_names) > 1, TRUE)
    expect_equal(length(error_var$value$multi_value$data) > 1, TRUE)
})


R/vars_manager.R
KB <- 1024
MB <- KB ** 2  # 1,048,576
GB <- KB ** 3  # 1,073,741,824
TB <- KB ** 4  # 1,099,511,627,776

DEFAULT_ABBREV_LEN <- 50

human_bytes <- function(B) {
    # Return the given bytes as a human friendly KB, MB, GB, or TB string
    if (B < KB) {
        return(sprintf("%.0f %s", B, `if`(B > 1, "Bytes", "Byte")))
    } else if (KB <= B & B < MB) {
        return(sprintf("%.2f KB", (B / KB)))
    } else if (MB <= B & B < GB) {
        return(sprintf("%.2f MB", (B / MB)))
    } else if (GB <= B & B < TB) {
        return(sprintf("%.2f GB", (B / GB)))
    } else if (TB <= B) {
        return(sprintf("%.2f TB", (B / TB)))
    }
}

make_multi_dict <- function(
    row_names, column_names, data, total_row_count=NULL, total_column_count=NULL
) {
    # column_count and row_count should be values for the whole data structure
    # column_names and row_names are values for the possibly abbreviated structure
    if (typeof(column_names) == "character") {
        column_names <- list(column_names)
    }
    if (typeof(row_names) == "character") {
        row_names <- list(row_names)
    }
    value <- list(
        column_count=`if`(!is.null(total_column_count), total_column_count, length(column_names)),
        row_count=`if`(!is.null(total_row_count), total_row_count, length(row_names)),
        column_names=column_names,
        row_names=row_names,
        data=data
    )
    return(list(multi_value=value))
}

get_single_vector_var <- function(obj, name) {
    return(list(
        summary=as.character(obj),
        abbrev=FALSE,
        value=list(single_value=as.character(obj))
    ))
}

get_vector_var <- function(obj, name, abbrev_len=DEFAULT_ABBREV_LEN) {
    summary <- sprintf("Length: %d", length(obj))
    abbrev <- !is.null(abbrev_len) && length(obj) > abbrev_len
    obj_pre <- `if`(abbrev, obj[1:abbrev_len], obj)
    data <- as.character(obj_pre)
    row_names <- list()
    i <- 1
    for (row_name in `if`(is.null(names(obj)), 1:length(obj_pre), names(obj_pre))) {
        row_names[i] = as.character(row_name)
        i <- i + 1
    }
    return(list(
        summary=summary,
        abbrev=abbrev,
        value=make_multi_dict(row_names, name, data, total_row_count=length(obj))
    ))
}

get_matrix_var <- function(obj, name, abbrev_len=DEFAULT_ABBREV_LEN) {
    dims <- dim(obj)
    summary <- sprintf("Size: %dx%d Memory: %s", dims[[1]], dims[[2]], human_bytes(utils::object.size(obj)))
    abbrev <- !is.null(abbrev_len) && length(obj) > abbrev_len
    obj_pre <- `if`(abbrev, obj[1:ceiling(abbrev_len/dims[[2]]),], obj)
    data <- list()
    for (i in 1:nrow(obj_pre)) {
        data[[i]] = as.list(as.character(obj_pre[i,]))
    }
    col_names <- list()
    i <- 1
    for (col_name in 1:ncol(obj_pre)) {
        col_names[i] <- as.character(col_name)
        i <- i + 1
    }
    row_names <- list()
    i <- 1
    for (row_name in 1:nrow(obj_pre)) {
        row_names[i] <- as.character(row_name)
        i <- i + 1
    }
    return(list(
        summary=summary,
        abbrev=abbrev,
        value=make_multi_dict(row_names, col_names, data, total_row_count=dims[[1]], total_column_count=dims[[2]])
    ))
}

get_dataframe_var <- function(obj, name, abbrev_len=DEFAULT_ABBREV_LEN) {
    dims <- dim(obj)
    summary <- sprintf("Size: %dx%d Memory: %s", dims[[1]], dims[[2]], human_bytes(utils::object.size(obj)))
    abbrev <- !is.null(abbrev_len) && ncol(obj)*nrow(obj) > abbrev_len
    obj_pre <- `if`(abbrev, obj[1:ceiling(abbrev_len/dims[[2]]),], obj)
    data <- list()
    if (length(obj) > 0) {
        for (i in 1:nrow(obj_pre)) {
            # If we use as.character on the whole row at once factors are represented as 
            # numbers and not their text value, so we as.character each cell
            data_row = list()
            for (j in 1:ncol(obj_pre)) {
                data_row[[j]] = as.character(obj_pre[i,j])
            }
            data[[i]] = data_row
        }
    }
    col_names <- list()
    if (ncol(obj_pre) > 0) {
        i <- 1
        for (col_name in `if`(length(names(obj_pre)) > 1, names(obj_pre), 1:ncol(obj_pre))) {
            col_names[i] <- sprintf("%s (%s)", as.character(col_name), toString(class(obj_pre[,i])))
            i <- i + 1
        }
    }
    row_names <- list()
    if (nrow(obj_pre) > 0) {
        i <- 1
        for (row_name in `if`(length(rownames(obj_pre)) > 1, rownames(obj_pre), 1:nrow(obj_pre))) {
            row_names[i] <- as.character(row_name)
            i <- i + 1
        }
    }
    return(list(
        summary=summary,
        abbrev=abbrev,
        value=make_multi_dict(row_names, col_names, data, total_row_count=dims[[1]], total_column_count=dims[[2]])
    ))
}

get_var_details <- function(obj, name, abbrev_len=DEFAULT_ABBREV_LEN) {
    # Get formatted info for a specific var. Pass abbrev_len=NULL
    # to get non-abbreviated variable info
    obj_type <- typeof(obj)
    obj_class <- class(obj)
    obj_length <- length(obj)

    intrinsic_types = c("character", "numeric", "integer", "logical", "complex", "raw")

    # vectors
    if (length(obj_class) == 1 && obj_class %in% intrinsic_types) {
        if (length(obj) == 1) {
            var_info <- get_single_vector_var(obj, name)
        } else {
            var_info <- get_vector_var(obj, name, abbrev_len)
        }
    } else if (is.data.frame(obj)) {
        var_info <- get_dataframe_var(obj, name, abbrev_len)
    } else if (is.list(obj)) {
        var_info <- get_vector_var(obj, name, abbrev_len)
    } else if (is.matrix(obj) && length(dim(obj)) == 2) {
        var_info <- get_matrix_var(obj, name, abbrev_len)
    } else {
        var_info = list(
            abbrev=FALSE,
            summary=noquote(toString(obj)),
            value=list(single_value=noquote(toString(obj)))
        )
    }

    return(list(
        name=name,
        type=toString(obj_class),
        abbreviated=var_info[["abbrev"]],
        summary=substr(var_info[["summary"]], 1, 140),  # limit summary length
        value=var_info[["value"]]
    ))
}

create_exception_var <- function(e) {
    stack <- sys.calls()
    row_names <- list()
    i <- 1
    for (row_name in `if`(is.null(names(stack)), 1:length(stack), names(stack))) {
        row_names[i] = as.character(row_name)
        i <- i + 1
    }
    return(
        list(
            name="Introspection Error",
            type=toString(class(e)),
            summary=toString(e),
            abbreviated=FALSE,
            value=list(
                multi_value=list(
                    column_count=1,
                    row_count=length(stack),
                    column_names=list("Traceback"),
                    row_names=row_names,
                    data=as.list(as.character(stack))
                )
            )
        )
    )
}

#' Get string containing json representation of defined vars in the provided environment
#'
#' @param envir An environment 
#' @param abbrev_len The length of elements at which vars should be abbreviated. Pass NULL to prevent abbreviation.
#'
#' @return A string with a json list of vars.
#' @export
#'
#' @examples
#' vars_json <- format_vars(environment())
format_vars <- function(envir, abbrev_len=DEFAULT_ABBREV_LEN) {
    # Get string containing json representation of currently defined vars
    # (pass the output of `environment()` to this function)
    current_vars <- list()

    err_resp <- tryCatch({
        idx <- 1
        for (name in ls(envir)) {
            obj <- get(name, envir=envir)
            if (startsWith(name, "__")) {
                next
            }
            obj_type <- typeof(obj)
            obj_class <- class(obj)
            if (grepl("closure", toString(obj_type)) || grepl("closure", toString(obj_class)) ) {
                next
            }
            var_details <- get_var_details(obj, name, abbrev_len)

            current_vars[[idx]] <- var_details
            idx <- idx + 1
        }
    }, error=function(e) {
        return(list(create_exception_var(e)))
    })
    return(rjson::toJSON(`if`(is.null(err_resp), current_vars, err_resp)))
}

#' Get string containing json representation of a single var in the provided environment
#'
#' @param envir An environment 
#' @param name Name of a variable in the environment
#' @param abbrev_len The length of elements at which vars should be abbreviated. Pass NULL to prevent abbreviation.
#'
#' @return A string with a json representation of a var.
#' @export
#'
#' @examples
#' x <- c(5,6)
#' vars_json <- format_var(environment(), "x", NULL)
format_var <- function(envir, name, abbrev_len=DEFAULT_ABBREV_LEN) {
    var_details <- NULL
    obj <- get(name, envir=envir)
    err_resp <- tryCatch({
        var_details <- get_var_details(obj, name, abbrev_len)
    }, error=function(e) {
        return(create_exception_var(e))
    })
    return(rjson::toJSON(`if`(is.null(err_resp), var_details, err_resp)))
}
